File: vault-service/internal/model/apierror.go
package model

// ApiError is the structured error envelope returned by Universal Auth services.
type ApiError struct {
	Code      string `json:"code"`
	Stage     string `json:"stage"`
	Message   string `json:"message"`
	TraceID   string `json:"trace_id,omitempty"`
	RequestID string `json:"request_id,omitempty"`
	Retryable bool   `json:"retryable"`
	Action    string `json:"action"`
}

func NewApiError(code, stage, message, action string, retryable bool) ApiError {
	return ApiError{
		Code:      code,
		Stage:     stage,
		Message:   message,
		Retryable: retryable,
		Action:    action,
	}
}

File: auth_client/app/src/main/java/com/aliddell/universalauth/data/ApiError.kt
package com.aliddell.universalauth.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ApiError(
    val code: String,
    val stage: String,
    val message: String,
    @SerialName("trace_id")
    val traceId: String? = null,
    @SerialName("request_id")
    val requestId: String? = null,
    val retryable: Boolean = false,
    val action: String = ""
) {
    fun toOperationError(): OperationError =
        OperationError(
            code = code,
            stage = stage,
            message = message,
            traceId = traceId,
            requestId = requestId,
            retryable = retryable,
            action = action
        )
}

class ApiException(
    val apiError: ApiError,
    message: String? = apiError.message
) : Exception(message)
